### PR TITLE
Display solution in exercise editor modal

### PR DESCRIPTION
**Pull request by vektor8891**
**Issue number:** #541 

**Changelog:**
 - Show preview of solution for single number in exercise editor modal

### CODE REVIEW

File: src/client/exercise/components/userControls/singleNumber/SingleNumberAdmin.tsx [hunk 1]
```diff
@@ -56,6 +60,14 @@ export function SingleNumberAdmin(bindProps: UseModelProps<UCSingleNumber>): JSX
             <div className="form-control-plaintext">
               Definiáld a megoldás függvényt:
               <code>(def solution-{data.name} #(...))</code>
+              <SingleNumber
+                disabled={true}
+                readonly={true}
+                ctrl={data}
+                onChange={noop}
+                name={data.name}
+                value={data.isDynamic ? evalPL(`(solution-${data.name})`) : data.solution}
```
Comment (maxinteger) [inline src/client/exercise/components/userControls/singleNumber/SingleNumberAdmin.tsx:69]: you do not need `data.isDynamic ? ` here since it was checked in line 59 and it must be `true` here 
also `value` prop accept only string value, so you have to convert the result into string

```value={evalPL(`(solution-${data.name})`)['toString']()}```

Note: 
the square brackets is needed for the `['toString']` because the `evalPL` returns with `unknown` type
it would be safer to check if the return value has `toString` method, but it is fine now

**Review by maxinteger — CHANGES_REQUESTED** [review]

**Review by vektor8891 — COMMENTED** [review]
@maxinteger could you pls take a look at the below question too?

File: src/client/exercise/components/userControls/singleNumber/SingleNumberAdmin.tsx (hunk 1 above)
Comment (vektor8891) [reply src/client/exercise/components/userControls/singleNumber/SingleNumberAdmin.tsx]: @maxinteger I've modified the code as you suggested. However, when I want to edit the field after saving, it gives me an error: `Uncaught TypeError: Cannot read property 'toString' of undefined` So it seems we do need to check if the return value has `toString` method - but how can I do that? When I simply try to use `try...catch` within the `<div>` I get an error: `expression expected ts(1109)`

Comment (maxinteger) [reply src/client/exercise/components/userControls/singleNumber/SingleNumberAdmin.tsx]: yes, you can not use statement in JSX (html in in the code) only expressions.

the `evalPL` returns with the return value of the `solutution-xyz` function which is defined in the exercise
It is return with `undefined` only when the function is not defined 

To avoid the problem you can do this:
`value={evalPL(`(solution-${data.name})`)?.['toString']() ?? ''}`
Where
-  `?.` mean, JS will check if the `evalPL` returns with `undefined` then it will not try to call the `toString`
- `??` mean if the left side was `undefined` then return the value from the right side (empty string) otherwise return the left side value

Basically we just make the expression safe without exception handling

Comment (vektor8891) [reply src/client/exercise/components/userControls/singleNumber/SingleNumberAdmin.tsx]: @maxinteger Thanks! It's working now.

Comment (vektor8891) [reply src/client/exercise/components/userControls/singleNumber/SingleNumberAdmin.tsx]: @maxinteger Actually I still get an error: `Object is of type 'unknown'.ts(2571)` Is there anything I can do to address this or we can just let it go?

Comment (maxinteger) [reply src/client/exercise/components/userControls/singleNumber/SingleNumberAdmin.tsx]: unfortunately you must cast the return value of `evalPL` with the `as` keyword:
```
  let solution = data.solution
  if (data.isDynamic) {
    const dynamicSolution = evalPL(`(solution-${data.name})`) as { toString(): string }
    if (dynamicSolution !== undefined) {
      solution = dynamicSolution.toString()
    }
  }
```
now you can use the `solution` variable when you render the `SingleNumber` component

Comment (vektor8891) [reply src/client/exercise/components/userControls/singleNumber/SingleNumberAdmin.tsx]: @maxinteger Thanks for the suggestion! It works now. I also added another feature to hide the hint if the dynamic solution is defined. Please let me know if there is anything else I need to do.

**Review by maxinteger — APPROVED** [review]
